### PR TITLE
Exclude hidden files by default

### PR DESCRIPTION
**Pull request by joshmgross**
Depends on https://github.com/actions/toolkit/pull/1791

Currently, all files within the search `path` are uploaded into the artifact. This includes hidden files, which could contain sensitive values that should not be accessible outside of the workflow run. To enhance the default security of this action, this PR excludes hidden files from the `path` by default. Users who require hidden files have validated that there are no sensitive values in their artifacts can opt-in to uploading hidden files with the `include-hidden-files` input.

This is a breaking change to the `v4` artifact that we will communicate out broadly before updating that major version tag.

### CODE REVIEW

File: merge/action.yml [hunk 1]
```diff
@@ -36,6 +36,11 @@ inputs:
       If true, the artifacts that were merged will be deleted.
       If false, the artifacts will still exist.
     default: 'false'
+  include-hidden-files:
```
Comment (joshmgross) [inline merge/action.yml:39]: Does it make sense for the `merge` sub-action to have this input too?

File: src/shared/search.ts [hunk 2]
```diff
@@ -11,11 +11,12 @@ export interface SearchResult {
...
 }
 
-function getDefaultGlobOptions(): glob.GlobOptions {
+function getDefaultGlobOptions(_includeHiddenFiles: boolean): glob.GlobOptions {
   return {
     followSymbolicLinks: true,
     implicitDescendants: true,
-    omitBrokenSymbolicLinks: true
+    omitBrokenSymbolicLinks: true,
+    // excludeHiddenFiles: !includeHiddenFiles,
```
Comment (joshmgross) [inline src/shared/search.ts:19]: Commented out until `@actions/glob` is released with this option

File: action.yml [hunk 3]
```diff
@@ -40,6 +40,11 @@ inputs:
       If false, the action will fail if an artifact for the given name already exists.
       Does not fail if the artifact does not exist.
     default: 'false'
+  include-hidden-files:
+    description: >
+      If true, hidden files will be included in the merged artifact.
+      If false, hidden files will be excluded from the merged artifact.
```
Comment (konradpabjan) [inline action.yml:46]: What about just `uploaded` vs `merged`?

The main action is different from the other merge action

File: README.md [hunk 4]
```diff
@@ -64,6 +64,7 @@ There is also a new sub-action, `actions/upload-artifact/merge`. For more info,
     Due to how Artifacts are created in this new version, it is no longer possible to upload to the same named Artifact multiple times. You must either split the uploads into multiple Artifacts with different names, or only upload once. Otherwise you _will_ encounter an error.
 
 3. Limit of Artifacts for an individual job. Each job in a workflow run now has a limit of 500 artifacts.
+4. With `v4.4` and later, hidden files are excluded by default.
```
Comment (reemasardana) [inline README.md:67]: Why are we introducing this for 4.4+ only?

File: merge/action.yml (hunk 1 above)
Comment (konradpabjan) [reply merge/action.yml:39]: More options are a good thing so I would steer towards having it available too.

**Review by konradpabjan — DISMISSED** [review]

File: README.md (hunk 4 above)
Comment (konradpabjan) [reply README.md:67]: This section of the `README` corresponds to breaking changes. We wrote this when there a ton of breaking changes between v3 and v4. With this new "default" behavior, we are introducing a new breaking change. Perhaps this can be expanded to clarify the behavior. What about something like:

```
With `v4.4` and later, hidden files are excluded by default. All prior versions of upload-artifact included hidden files by default
```

File: README.md [hunk 5]
```diff
@@ -107,6 +108,10 @@ For assistance with breaking changes, see [MIGRATION.md](docs/MIGRATION.md).
     # Does not fail if the artifact does not exist.
     # Optional. Default is 'false'
     overwrite:
+
+    # Whether to include hidden files in the provided path in the artifact
+    # Optional. Default is 'false'
+    include-hidden-files:
```
Comment (SrRyan) [inline README.md:114]: Do we want to put any warning within this comment?

File: action.yml (hunk 3 above)
Comment (joshmgross) [reply action.yml]: Kept it simple and removed `merged` - https://github.com/actions/upload-artifact/pull/598/commits/710f36207581d49e465ccbed3e007c317290fe11

File: README.md (hunk 5 above)
Comment (joshmgross) [reply README.md:116]: Added a warning - https://github.com/actions/upload-artifact/pull/598/commits/d52396ac5d312b1bda1b9fe2ae55d862c65abd68

File: docs/MIGRATION.md [hunk 6]
```diff
@@ -189,21 +190,59 @@ jobs:
...
 Note that this will download all artifacts to a temporary directory and reupload them as a single artifact. For more information on inputs and other use cases for `actions/upload-artifact/merge@v4`, see [the action documentation](../merge/README.md).
+
+## Hidden Files
+
+By default, hidden files are ignored by this action to avoid unintentionally uploading sensitive
+information.
+
+In versions of this action before v4.4.0, these hidden files were included by default.
+
+If you need to upload hidden files, you can use the `include-hidden-files` input.
```
Comment (SrRyan) [inline docs/MIGRATION.md:219]: Do we want to mention anything about wildcards and potentially the impact of using them?

**Review by SrRyan — APPROVED** [review]

File: docs/MIGRATION.md (hunk 6 above)
Comment (joshmgross) [reply docs/MIGRATION.md:219]: What's the impact of using wildcards?

File: src/shared/search.ts [hunk 7]
```diff
@@ -11,11 +11,12 @@ export interface SearchResult {
   rootDirectory: string
 }
 
-function getDefaultGlobOptions(): glob.GlobOptions {
+function getDefaultGlobOptions(includeHiddenFiles: boolean): glob.GlobOptions {
```
Comment (reemasardana) [inline src/shared/search.ts:14]: nit: Can we just call this excludeHiddenFiles like we do in the glob options?

File: docs/MIGRATION.md (hunk 6 above)
Comment (SrRyan) [reply docs/MIGRATION.md:219]: Let me rephrase to - "When using wildcards be sure that you aren't accidentally uploading hidden files that may contain secrets"

File: README.md (hunk 4 above)
Comment (crazy-max) [reply README.md:67]: What about GHES users where only v3 works (with actions/artifact lib v1)?

Where GHES is going to support v4 btw?

Comment (SrRyan) [reply README.md:67]: We are going to backport a change to v3 as there is no timeline currently for v4 to be in GHES.

**Review by reemasardana — APPROVED** [review]

File: dist/merge/index.js [hunk 8]
```diff
@@ -9104,6 +9121,10 @@ class DefaultGlobber {
                 if (!stats) {
                     continue;
                 }
+                // Hidden file or directory?
+                if (options.excludeHiddenFiles && path.basename(item.path).match(/^\./)) {
```
Comment (casperdcl) [inline dist/merge/index.js:9125]: what about gh-pages' `.nojekyll`?

File: dist/upload/index.js [hunk 9]
```diff
@@ -9104,6 +9121,10 @@ class DefaultGlobber {
                 if (!stats) {
                     continue;
                 }
+                // Hidden file or directory?
+                if (options.excludeHiddenFiles && path.basename(item.path).match(/^\./)) {
```
Comment (casperdcl) [inline dist/upload/index.js:9125]: same here

**Review by konradpabjan — APPROVED** [review]

File: dist/merge/index.js (hunk 8 above)
Comment (nebuk89) [reply dist/merge/index.js:9125]: Thank you for the feedback, for now we are going to go ahead with the 'all in' option as starting to pick and choose common things is a risk we start to go back to 'too many' common things and expose folks to the risk of mistakes causing issues. 
I get the idea though, maybe in the future we could look at 'more flags' for "all-common-none", but that will be something we need to revisit at a later date so we could get a better view on what common would be :)

Comment (casperdcl) [reply dist/merge/index.js:9125]: Well at least there should probably be a black/whitelist option rather than a `bool` methinks

**Review by robherley — APPROVED** [review]

File: src/shared/search.ts (hunk 7 above)
Comment (joshmgross) [reply src/shared/search.ts:14]: Going to leave this as-is since we've back-ported the changes with this variable name. We can fix it in a follow-up if it gets confusing.

File: dist/merge/index.js (hunk 8 above)
Comment (joshmgross) [reply dist/merge/index.js:9125]: `path` is effectively that allowlist/denylist option - you can allow hidden files and then specify in `path` which to exclude

https://github.com/actions/upload-artifact#upload-using-multiple-paths-and-exclusions

File: dist/upload/index.js (hunk 9 above)
Comment (joshmgross) [reply dist/upload/index.js:9125]: See https://github.com/actions/upload-artifact/pull/598#discussion_r1729295064

**Comment by bartekpacia** [comment]
This just broke my workflow. I think this warrants a breaking change.

**Comment by AlexSkrypnyk** [comment]
This has just broke 1000s of workflows.

**Comment by neoye911** [comment]
> This has just broke 1000s of workflows that rely on artifacts produced by tools like pcov/phpunit etc. and any other reporting tools.
> 
> This should have been released as a major version.

Failed our release pipeline as well. Agree.

**Comment by thomas-alias-zaiko** [comment]
Hey guys, even though this is definitely a good change, introducing a breaking change on a minor version isn't great.
Like the above comments, this also affected our company's workflows (including our production ones).
Please consider breaking changes on major versions next time 🙂

**Comment by casperdcl** [comment]
It's also likely extremely hard to debug for users. "Breaking" was included in #602's title so it's unforgivable to sneak this in a minor version.

**Comment by twm** [comment]
This change likely breaks every workflow that uploads the `.coverage*` files generated by [coverage.py](https://coverage.readthedocs.io/en/latest/). This will have a significant impact in the Python ecosystem.

Did GitHub assess the number of (public) artifact uploads that would be affected by this change? I understand that avoiding upload of `.git/config` files containing credentials was the goal. Is forbidding the upload of any pathname beginning with `.` the lowest-impact way to achieve that aim?

Some other possibilities:

1. [Don't upload `.git` directories by default](https://github.com/actions/upload-artifact/pull/599).
2. Don't put credentials in `.git/config` by default.

The first option seems attractive to me, regardless of security implication, since the `.git` directory is redundant with the content of the Git repository. I struggle to envision a scenario where I would deliberately upload the `.git` directory.

The second option minimizes harm.

I must express my sympathy with all those who advocated for less impactful options, but did not prevail. Sadly, the OSS ecosystem must deal with the consequences.

**Comment by casperdcl** [comment]
I'd still suggest reverting in a minor version and re-releasing in a major one.

**Comment by bartekpacia** [comment]
Yeah, revert this ASAP to not require waste thousands of hours by many people who all will have to investigate this now.

**Comment by joshmgross** [comment]
Please direct all feedback to the tracking issue https://github.com/actions/upload-artifact/issues/602
